### PR TITLE
ref: Move build configuration from `setup.py` to `pyproject.toml`; apply various repo/CI refactors

### PR DESCRIPTION
**Pull request by lgarber-akamai**
## 📝 Description

This change makes the following build system improvements to the Linode CLI:

* Moving away from `setup.py` in favor of setuptools in `pyproject.toml`
* Moving away from the `baked_version` construct in favor of injecting the version into a new `version.py` file.
* Removing `requirements.txt` and `requirements-dev.txt` in favor of `pyproject.toml`
* Removing the logic to automatically bake and install bash completions to `/etc/bash_completion.d`
    * This was discussed internally as a team
* Removing the Jenkinsfile (no longer used)
* Updating GHA workflows and Dockerfiles accordingly
* Updating the OCI publish workflow to not rely on the `version` script (which was previously removed)

## ✔️ How to Test

The following test steps assume you have pulled down this PR locally.

### Testing Workflows

For convenience I tested all workflows on the `main` branch of my fork; they can be viewed here:
* [OCI Build & Publish](https://github.com/lgarber-akamai/linode-cli/actions/runs/8804593829/job/24165298002)
* [Release](https://github.com/lgarber-akamai/linode-cli/actions/runs/8804593828/job/24165298009)
* [Integration Tests (failing due to unrelated configuration issue)](https://github.com/lgarber-akamai/linode-cli/actions/runs/8804591132/job/24165287589) 
* [PR Integration Tests (failing due to unrelated issues)](https://github.com/lgarber-akamai/linode-cli/actions/runs/8804876351)

### Testing the Build & Install Process

1. Run the following command to build & install the CLI:

```
make install
```

2. Ensure the CLI has installed successfully and the installed version matches `v0.0.0.dev`:

```
linode-cli --version

# linode-cli 0.0.0.dev0
# Built from spec version 4.174.0
```

3. Install the CLI with an overridden version:

```
LINODE_CLI_VERSION=1.2.3 make install
```

4. Ensure the CLI has installed correctly and the installed version matches `v0.0.0.dev`:

```
linode-cli --version

# linode-cli 1.2.3
# Built from spec version 4.174.0
```

### Testing the Dockerfile Changes

1. Create a read-only GitHub Personal Access Token and export it under the `GITHUB_TOKEN` environment variable.

2. Run the following command to build the CLI OCI image:

```
docker build -t local-linode-cli --build-arg linode_cli_version=1.2.3 --build-arg github_token=$GITHUB_TOKEN .
```

3. Attempt to execute a CLI command, for example:

```
docker run --rm -it -e LINODE_CLI_TOKEN=$LINODE_TOKEN local-linode-cli linodes ls
```

4. Ensure the command runs as expected.

### Unit Testing

```
make testunit
```

### Integration Testing

```
make testint
```

### CODE REVIEW

File: linodecli/__init__.py [hunk 1]
```diff
@@ -112,7 +109,7 @@ def main():  # pylint: disable=too-many-branches,too-many-statements
         if not parsed.command:
             # print version info and exit - but only if no command was given
             print(f"linode-cli {VERSION}")
-            print(f"Built off spec version {cli.spec_version}")
+            print(f"Built from spec version {cli.spec_version}")
```
Comment (lgarber-akamai) [inline linodecli/__init__.py:112]: This isn't really relevant to the rest of these changes, but the wording was bothering me a bit 🙂

File: linodecli/__init__.py [hunk 2]
```diff
@@ -21,17 +21,14 @@
...
 from .helpers import handle_url_overrides
 from .output import OutputMode
+from .version import __version__
 
-# this might not be installed at the time of building
-try:
-    VERSION = version("linode-cli")
-except:
-    VERSION = "building"
+VERSION = __version__
```
Comment (lgarber-akamai) [inline linodecli/__init__.py:31]: This should be much more reliable than grabbing the version from the package itself.

File: .github/workflows/publish-oci.yml [hunk 3]
```diff
@@ -34,13 +34,18 @@ jobs:
       # This is necessary as we want to ensure that version tags
       # are properly formatted before passing them into the
       # DockerFile.
-      - name: Get CLI version
```
Comment (lgarber-akamai) [inline .github/workflows/publish-oci.yml:37]: I honestly have no clue how this worked in the last release given we removed the `version` script beforehand. Regardless, I moved this over to a more reliable solution 🙂

File: pyproject.toml [hunk 4]
```diff
@@ -2,6 +2,51 @@
 requires = ["setuptools", "wheel", "packaging"]
 build-backend = "setuptools.build_meta"
 
+[project]
+name = "linode-cli"
+authors = [{ name = "Akamai Technologies Inc.", email = "developers@linode.com" }]
```
Comment (lgarber-akamai) [inline pyproject.toml:7]: Just a note, I swapped this from `Linode` to `Akamai Technologies Inc.`

File: Makefile [hunk 5]
```diff
@@ -27,13 +32,17 @@ else
 	cp data-3 linodecli/
 endif
 
+.PHONY: create-version
+create-version:
+	@printf "${VERSION_MODULE_DOCSTRING}__version__ = \"${LINODE_CLI_VERSION}\"\n" > $(VERSION_FILE)
+
 .PHONY: build
-build: clean bake
+build: clean  create-version bake
```
Comment (zliang-akamai) [inline Makefile:40]:
```suggestion
build: clean create-version bake
```

File: tests/unit/test_completion.py [hunk 6]
```diff
@@ -3,7 +3,7 @@
 Unit tests for linodecli.completion
 """
 
-from unittest.mock import mock_open, patch
+pass
```
Comment (zliang-akamai) [inline tests/unit/test_completion.py:6]: Let's just remove it?

**Review by zliang-akamai — APPROVED** [review]
Nice work! The code base is much cleaner now. 🎉

File: tests/unit/test_completion.py (hunk 6 above)
Comment (lgarber-akamai) [reply tests/unit/test_completion.py:6]: Oops, no clue how that got there lol. Let me fix that up

**Review by ykim-1 — APPROVED** [review]
Nice refactor! LGTM

**Pull request merged**